### ISSUE TITLE
confusing line numbers in feedback from TestCaseGui

### ISSUE THREAD

**Issue by presnick** [RunestoneInteractive/RunestoneComponents#222]
As you can see here, the error message refers to line number 10, which is a hidden line of code containing the Unit Test. Any thoughts on how we could provide something more understandable? I think we could have related problems when there is prefix code: an error message referring to a line number in the student's code will have the wrong line number.

![image](https://cloud.githubusercontent.com/assets/1077198/17601996/b21f0fe0-5fd8-11e6-8d93-7765a028dc20.png)

Here's the source code:

.. activecode:: ee_ch9_01
   
   lst = ["hi", "morning", "dog", "506", "caterpillar", "balloons", 106, "yo-yo", "python", "moon", "water", "sleepy", "daffy", 45, "donald", "whiteboard", "glasses", "markers", "couches", "butterfly", "100", "magazine", "door", "picture", "window", ["Olympics", "handle"], "chair", "pages", "readings", "burger", "juggle", "craft", ["store", "poster", "board"], "laptop", "computer", "plates", "hotdog", "salad", "backpack", "zipper", "ring", "watch", "finger", "bags", "boxes", "pods", "peas", "apples", "horse", "guinea pig", "bowl", "EECS"]
   
   =====

   from unittest.gui import TestCaseGui

   class myTests(TestCaseGui):

      def testOne(self):
         self.assertEqual(output, "laptop", "Testing that output value is assigned to correct value.")

   myTests().main()

**Comment by bnmnetp** [comment]
@millro04 Any thoughts on how we might improve this?

**Comment by millro04** [comment]
Do we have any control over the error message "Error: NameError: name 'output' is not defined on line 10"?

**Comment by bnmnetp** [comment]
In gui.py main we capture the exception.  So we can get the error message out of the exception and do something with it.

**Comment by millro04** [comment]
Are most of the exception messages confusing like this one? If so....we could always still catch the exception, but send back a more generic message - "there was a runtime error", "An exception occurred", etc. without giving specifics that will probably just confuse them anyways. And we still have the custom message from the assert equal statement appear when there is no exception. So that would be the message that isn't as generic, since an actual person creates that message and not the cpu.

**Comment by bnmnetp** [comment]
@presnick what are you looking for here?

**Comment by presnick** [comment]
Best would be to have a line number that corresponds to the line numbers
shown in the activecode window.

Could you have the line numbers shown in the activecode window have an
offset based on the number of lines of "pre" code? That way, we could leave
the error messages as is. Sometimes line numbers wouldn't start at 1, but I
think that would be OK for students.

**Comment by bnmnetp** [comment]
But, that would not help in this case because the error occurs in the unittest code.  In this example the student is suposed to simply set a variable to a value, but they don't set the right variable.  the error happens when the assert statement uses a variable that has not been defined.

**Comment by bnmnetp** [comment]
IF we had the source for the offending line, then maybe the best thing to do would be to display the error message without the line number, and also show the code from the line that failed.  

The difficult part here is that we can only do so much,  as the unittest module is part of skulpt and  not part of the activecode component.  We are just trying to do our best with the output from the standard unittest module.

**Comment by presnick** [comment]
Good point. Sorry, I got confused about this vs. problems with pre-code. (I still think when there's pre-code, we should start our line numbers after the line numbers for the hidden code.)

Back to this problem. I presume that once we get the error message back from skulpt unittest module, extract the line number from the error message and check whether that line number matches a line number that is currently displayed. If so, leave the error message as is. If not, remove the phrase "on line n". 

I don't think you should show the code for the hidden line that failed. That will be confusing, and may also reveal something that we'd rather not reveal to students (e.g., in unittests on a timed exam).

**Comment by bnmnetp** [comment]
OK, I'll buy starting the line numbers properly if there is prefix code.  We should make that a separate issue to track, as that should be fully under the control of the activecode component.

Given where we are in the summer, I have classes starting in 13 days -- and I suppose I should prep for that.  We are going to have to really use the github issues mechanism and really prioritize what we work on. There is a lot of cleanup to do on the new interface, and I know from past years that we won't really start getting feedback until classes start.

**Comment by presnick** [comment]
Sounds like a good solution for now; always remove "on line n". Will add the other issue.
